### PR TITLE
Avoid using process.env.NODE_ENV direcly

### LINKED ISSUE #1168: Avoid using process.env.NODE_ENV
There's no such thing in a browser

### PR DESCRIPTION
**Pull request by kamilkisiela**
Fixes #1168

TODO:

- [x] If this PR is a new feature, reference an issue where a consensus about the design was reached (not necessary for small changes)
- [x] Make sure all of the significant new logic is covered by tests
- [x] Rebase your changes on master so that they can be merged easily
- [x] Make sure all tests and linter rules pass
- [x] Update CHANGELOG.md with your change
- [x] Add your name and email to the AUTHORS file (optional)
- [x] If this was a change that affects the external API, update the docs and post a link to the PR in the discussion

### CODE REVIEW

File: src/util/environment.ts [hunk 1]
```diff
@@ -0,0 +1,15 @@
+export function isEnv(env: string): boolean {
```
Comment (DxCx) [inline src/util/environment.ts]: that's abit tricky.
if process is not define, all the checks will return false.
so you are not production not development and not test?
we need to define default behavior, @helfer / @stubailo which one do you think?
and then make sure that at least 1 will return true in that case.

Comment (kamilkisiela) [inline src/util/environment.ts]: We don't use the default behavior, in every conditional statement there is available we check for specific environment so I decided to just look just for specific env and that's it.

If there's no `process`:
`isProduction()` returns false
`isTest()` returns false
`isDevelopment()` returns false too

Comment (kamilkisiela) [inline src/util/environment.ts]: [this line](https://github.com/apollostack/apollo-client/pull/1169/files#diff-658f6fae1df103e086034713f52c5672L228):

```ts
typeof process === 'undefined' || (process.env && process.env.NODE_ENV !== 'production')
```

could be replaced with `!isProduction()` and every case seems give the same result.

`!isProduction()` when :
- `process = undefined` gives `true`
- `process.env.NODE_ENV = 'test'` gives `true`
- `process.env.NODE_ENV = 'development'` gives `true`
- `process.env.NODE_ENV = 'production'` gives `false`

Comment (kamilkisiela) [inline src/util/environment.ts]: But if you want to have the default behavior, let's say `development` if there's no process, I could replace `isDevelopment()` with this:

```ts
return isEnv('development') || typeof process === 'undefined';
```

this way we make `isEnv()` to strictly check if there's a value and `isDevelopment()` to get our goal, which is a default behavior.

@DxCx what do you think?

Comment (DxCx) [inline src/util/environment.ts]: alright, sounds good to me.

**Comment by Poincare** [comment]
@stubailo I think you mentioned that `NODE_ENV` is available on browsers due to the package manager or something; could you explain that further?

**Comment by kamilkisiela** [comment]
@Poincare https://webpack.js.org/guides/production-build

**Comment by helfer** [comment]
@kamilkisiela I like the PR, but I think we should make one of them be true when we can't figure out what environment it is. When in doubt we should probably assume that it's a production environment, because these guards are usually there to make sure that we don't do something costly or noisy in a production environment.

**Comment by Poincare** [comment]
Maybe we should print a warning if we can't figure out the environment. I would be very puzzled if the network tab suddenly started showing only `{id: 18}` instead of the printed query because of a shell misconfiguration or something.

**Comment by kamilkisiela** [comment]
@helfer I made it to use `production` as the default environment but... it's a little breaking change in terms of DevTools. 

Before that, we assumed that if there's no `process` we set `defaultConnectToDevTools = true` to attach the client instance to window.

Now we treat no process as it'd be in production. So we force users to define `process.env.NODE_ENV = 'development'`.

We could make an exception and set `defaultConnectToDevTools` to `true` if there's no process. So if someone has `process.env.NODE_ENV` properly defined, it'll work.

**Comment by stubailo** [comment]
Yeah I think we definitely want dev tools on by default, otherwise we'll need to build better handling into the tools to print something nice when they can't find Apollo client, which would be a lot more work. Although, probably a better solution long-term.

**Comment by stubailo** [comment]
Given that the main use of prod vs. dev is to print better errors, use safer behavior, etc, I would say having dev on by default is better unless we can reliably notify developers to set dev vs. prod.

File: src/util/maybeDeepFreeze.ts [hunk 2]
```diff
@@ -16,7 +20,7 @@ function deepFreeze (o: any) {
 };
 
 export default function maybeDeepFreeze(obj: any) {
-  if (process.env.NODE_ENV === 'development' || process.env.NODE_ENV === 'test') {
+  if (isDevelopment() || isTest()) {
```
Comment (kamilkisiela) [inline src/util/maybeDeepFreeze.ts]: Maybe we should change it to `!isProduction()`, or you want to leave it as is, so only for the default env (development) and test?

**Comment by calebmer** [comment]
A little late to the party, but do you think it would also be useful to check for a global `__DEV__` variable? This is what is used in React Native to switch between development and production modes.

**Comment by helfer** [comment]
Thanks @kamilkisiela !

@calebmer I'll merge this now to unblock Uri, but I'd be in favor of amending the code to use `__DEV__` if that's a pretty standard thing. Otherwise we can also just leave it as is.

**Pull request merged**